crates/storage/api/mod.rs

```diff
@@ -0,0 +1,106 @@
+//! # Storage Backend API
+//!
+//! This module provides a thin, minimal interface for storage backends:
+//!
+//! - Thin: Minimal set of operations that databases must provide
+//! - Simple: Avoids type-system complexity and focuses on core functionality
+//!
+//! Rather than implementing business logic in each database backend, this API
+//! provides low-level primitives that higher-level code can build upon.
+//! This eliminates code duplication and makes adding new database backends trivial.
+//!
+//! The API differentiates between three types of database access:
+//!
+//! - Read transactions ([`StorageReadTx`]): Read-only views of the database,
+//!   with no atomicity guarantees between operations.
+//! - Write transactions ([`StorageWriteTx`]): Write batch functionality, with
+//!   atomicity guarantees at commit time.
+//! - Locked snapshots ([`StorageLocked`]): Read-only views of a point in time, right now it's
+//!   only used in snapsync stage.
```

Comment:
IMO this should be how most things during block execution should work.

Reply:
Ah, but that's the opposite of what I'm implying. I think tries should be created based on static snapshots, rather than allowing changes under one's feet. How we implement that is indifferent, but I wouldn't want to deal with trie instance invalidation during e.g. iteration.